aea/configurations/data_types.py

```diff
@@ -0,0 +1,827 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""Base config data types."""
+import functools
+import re
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import (
+    Collection,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import semver
+from packaging.specifiers import SpecifierSet
+from urllib3.util import Url, parse_url
+
+from aea.configurations.constants import (
+    AGENT,
+    CONNECTION,
+    CONTRACT,
+    DEFAULT_GIT_REF,
+    PROTOCOL,
+    SKILL,
+)
+from aea.exceptions import enforce
+from aea.helpers.base import (
+    RegexConstrainedString,
+    STRING_LENGTH_LIMIT,
+    SimpleId,
+    SimpleIdOrStr,
+)
+
+
+T = TypeVar("T")
+
+
+VersionInfoClass = semver.VersionInfo
+PackageVersionLike = Union[str, semver.VersionInfo]
+
+
+class JSONSerializable(ABC):
+    """Interface for JSON-serializable objects."""
+
+    @property
+    @abstractmethod
+    def json(self) -> Dict:
+        """Compute the JSON representation."""
+
+    @classmethod
+    def from_json(cls, obj: Dict):
+        """Build from a JSON object."""
```

Comment:
One day, instead of this, we could have external JSON translators. One way to achieve this is to use [`singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch) like:

```python

def to_json(config):
    raise NotImplementedError

@to_json.register(config: AgentConfig):
def _(config):
    ....

@to_json.register(config: ProtocolConfig):
def _(config):
    ....

# ... the same for from_json, but dispatching on the type instead.




```

Reply:
That would achieve single-responsibility for configuration classes (i.e. config classes should only care about information that they carry, and not how to serialize-deserialize).

Reply:
I think it's loader related, but, yes, we can update as well.